### PR TITLE
Blackbox regressor support + shadow support for additional models

### LINKED ISSUE #1782: Blackbox regression support
**Is your feature request related to a problem? Please describe.**
At the moment regression model support is limited to a few frameworks (pytorch, sklearn, etc.)

**Describe the solution you'd like**
Similar to the BlackBoxClassifier it would be beneficial to also have a BlackBoxRegressor (for when the model object is not available or the correct framework-specific wrapper is not available).

### PR DESCRIPTION
**Pull request by abigailgold**
# Description

Added BlackBox regressor wrapper + the ability to export it as a classifier if supplying threshold cutoff values.
Implemented clone_for_refitting for KerasEstimator and XGBoostClassifier and updated the list of supported estimators in ShadowModel.

Fixes #1782

## Type of change

Please check all relevant options.

- [x] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

Added new tests for BlackboxRegressor and XGBoost.

**Test Configuration**:
- OS: MacOS 12.6
- Python version: 3.8

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: tests/estimators/regression/test_blackbox.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
+# MIT License
+#
+# Copyright (C) The Adversarial Robustness Toolbox (ART) Authors 2021
```
Comment (beat-buesser) [inline tests/estimators/regression/test_blackbox.py:3]:
```suggestion
# Copyright (C) The Adversarial Robustness Toolbox (ART) Authors 2022
```

File: art/estimators/keras.py [hunk 2]
```diff
@@ -83,3 +88,49 @@ def compute_loss(self, x: np.ndarray, y: np.ndarray, **kwargs) -> np.ndarray:
         :rtype: Format as expected by the `model`
         """
         raise NotImplementedError
+
+    def clone_for_refitting(
+        self,
+    ) -> "KERAS_ESTIMATOR_TYPE":  # lgtm [py/inheritance/incorrect-overridden-signature]
```
Comment (beat-buesser) [inline art/estimators/keras.py:94]:
```suggestion
    ) -> "KERAS_ESTIMATOR_TYPE":
```

File: art/estimators/keras.py [hunk 3]
```diff
@@ -83,3 +88,49 @@ def compute_loss(self, x: np.ndarray, y: np.ndarray, **kwargs) -> np.ndarray:
...
+        """
+        Create a copy of the estimator that can be refit from scratch. Will inherit same architecture, optimizer and
+        initialization as cloned model, but without weights.
+
+        :return: new estimator
+        """
+
+        import tensorflow as tf  # lgtm [py/repeated-import]
+        import keras  # lgtm [py/repeated-import]
```
Comment (beat-buesser) [inline art/estimators/keras.py:103]:
```suggestion
        import tensorflow as tf
        import keras
```

File: art/estimators/classification/xgboost.py [hunk 4]
```diff
@@ -152,29 +161,48 @@ def predict(self, x: np.ndarray, **kwargs) -> np.ndarray:
...
+    def clone_for_refitting(
+        self,
+    ) -> "XGBoostClassifier":  # lgtm [py/inheritance/incorrect-overridden-signature]
+        """
+        Create a copy of the estimator that can be refit from scratch. Only supported for models of type XGBClassifier.
+
+        :return: new estimator.
+        :raises `NotImplementedError`: This method is not supported for Booster objects.
+        """
+        import xgboost  # lgtm [py/repeated-import] lgtm [py/import-and-import-from]
```
Comment (beat-buesser) [inline art/estimators/classification/xgboost.py:173]:
```suggestion
        import xgboost
```

File: art/estimators/classification/xgboost.py [hunk 5]
```diff
@@ -152,29 +161,48 @@ def predict(self, x: np.ndarray, **kwargs) -> np.ndarray:
 
         return y_prediction
 
-    def _get_nb_classes(self, nb_classes: Optional[int]) -> int:
+    def clone_for_refitting(
+        self,
+    ) -> "XGBoostClassifier":  # lgtm [py/inheritance/incorrect-overridden-signature]
```
Comment (beat-buesser) [inline art/estimators/classification/xgboost.py:166]:
```suggestion
    ) -> "XGBoostClassifier":
```

File: art/estimators/classification/xgboost.py [hunk 6]
```diff
@@ -114,16 +116,23 @@ def nb_features(self) -> int:
...
         :param y: Target values (class labels) one-hot-encoded of shape (nb_samples, nb_classes).
         :param kwargs: Dictionary of framework-specific arguments. These should be parameters supported by the
-                       `fit` function in `xgboost.Booster` or `xgboost.XGBClassifier` and will be passed to this
-                       function as such.
-        :raises `NotImplementedException`: This method is not supported for XGBoost classifiers.
+                       `fit` function in `xgboost.XGBClassifier` and will be passed to this function as such.
+        :raises `NotImplementedException`: This method is not supported for Booster objects.
         """
-        raise NotImplementedError
+        import xgboost  # lgtm [py/repeated-import] lgtm [py/import-and-import-from]
```
Comment (beat-buesser) [inline art/estimators/classification/xgboost.py:127]:
```suggestion
        import xgboost
```

File: art/attacks/inference/membership_inference/shadow_models.py [hunk 7]
```diff
@@ -24,14 +24,27 @@
 
 import math
 from functools import reduce
-from typing import Callable, Tuple, TYPE_CHECKING, Union, List, Optional, Sequence
+from typing import Callable, Tuple, TYPE_CHECKING, Union, List, Sequence, Optional
```
Comment (beat-buesser) [inline art/attacks/inference/membership_inference/shadow_models.py:27]: Is this change required?

File: art/estimators/regression/blackbox.py [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+
+logger = logging.getLogger(__name__)
+
+
+class BlackBoxRegressor(RegressorMixin, BaseEstimator):
+    """
+    Class for black-box regressors.
+    """
+
+    estimator_params = Regressor.estimator_params + ["input_shape" + "predict_fn" + "loss_fn"]
```
Comment (beat-buesser) [inline art/estimators/regression/blackbox.py:46]:
```suggestion
    estimator_params = Regressor.estimator_params + ["input_shape", "predict_fn", "loss_fn"]
```

File: tests/estimators/regression/test_blackbox.py [hunk 9]
```diff
@@ -0,0 +1,114 @@
+# MIT License
```
Comment (beat-buesser) [inline tests/estimators/regression/test_blackbox.py:1]: I think to run these tests the file `run_tests.sh` needs to be updated with
```python
pytest --cov-report=xml --cov=art --cov-append  -q -vv tests/estimators/regression/ --framework=$framework  --durations=0
    if [[ $? -ne 0 ]]; then exit_code=1; echo "Failed estimators/regression tests"; fi
```
starting in line 70.

**Comment by abigailgold** [comment]
> Should we also move `TensorFlowV2Classifier.clone_for_refitting` to `TensorFlowV2Estimator` similar as for Keras? Maybe also for PyTorch?

In both PyTorchClassifier and TensorFlowV2Classifier, nb_classes must be provided to the init method. So the code in these cases for classifiers and regressors is different. In Keras the code was identical.

**Review by beat-buesser — APPROVED** [review]
Hi @abigailgold Thank you very much for contributing black-box support for regression models to ART! Looks good to me.
